contracts/src/contracts/ValidatorManager/BalancerValidatorManager.sol

```diff
@@ -0,0 +1,320 @@
+        uint64 initialSecurityModuleWeight
+    ) external initializer {
+        __BalancerValidatorManager_init(
+            settings, initialOwner, initialSecurityModule, initialSecurityModuleWeight
+        );
+    }
+
+    function __BalancerValidatorManager_init(
+        ValidatorManagerSettings calldata settings,
+        address initialOwner,
+        address initialSecurityModule,
+        uint64 initialSecurityModuleWeight
+    ) internal onlyInitializing {
+        __ValidatorManager_init(settings);
+        __Ownable_init(initialOwner);
+        _setupSecurityModule(initialSecurityModule, initialSecurityModuleWeight);
+    }
+
+    // solhint-disable-next-line no-empty-blocks
+    function __BalancerValidatorManager_init_unchained() internal onlyInitializing {}
+
+    // solhint-enable func-name-mixedcase
+
+    /// @inheritdoc IBalancerValidatorManager
+    function setupSecurityModule(address securityModule, uint64 maxWeight) external onlyOwner {
+        _setupSecurityModule(securityModule, maxWeight);
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function initializeValidatorRegistration(
+        ValidatorRegistrationInput calldata registrationInput,
+        uint64 weight
+    ) external onlySecurityModule returns (bytes32 validationID) {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+
+        if (!$.securityModules.contains(msg.sender)) {
+            revert BalancerValidatorManager__SecurityModuleNotRegistered(msg.sender);
+        }
+
+        validationID = _initializeValidatorRegistration(registrationInput, weight);
+
+        // Update the security module weight
+        uint64 newSecurityModuleWeight = $.securityModuleWeight[msg.sender] + weight;
+        _updateSecurityModuleWeight(msg.sender, newSecurityModuleWeight);
+
+        $.validatorSecurityModule[validationID] = msg.sender;
+    }
+
+    /// @inheritdoc IBalancerValidatorManager
+    function initializeEndValidation(
+        bytes32 validationID
+    ) external onlySecurityModule {
+        BalancerValidatorManagerStorage storage $ = _getBalancerValidatorManagerStorage();
+        Validator memory validator = getValidator(validationID);
+
+        // Ensure the validator weight is not being updated
+        if ($.validatorPendingWeightUpdate[validationID] != 0) {
+            revert BalancerValidatorManager__NoPendingWeightUpdate(validationID);
+        }
+
+        _checkValidatorSecurityModule(validationID, msg.sender);
+        _initializeEndValidation(validationID);
+
+        // If the validator is not an initial validator, update the security module weight
+        if ($.validatorSecurityModule[validationID] != address(0)) {
+            // Update the security module weight
+            uint64 newSecurityModuleWeight = $.securityModuleWeight[msg.sender] - validator.weight;
+            _updateSecurityModuleWeight(msg.sender, newSecurityModuleWeight);
+        }
```

Comment:
To avoid having different behaviors for initial validators, can't we generate a `validationID` from the convert tx id like they do [here](https://github.com/ava-labs/teleporter/blob/fa9d24a7fbb388e6316f19e9c78b44dbf12cd772/contracts/validator-manager/ValidatorManager.sol#L172)?

Reply:
Just realized we don't have the `ConversionData` here so that won't work...

Reply:
So a weight change or validation end to an initial validator won't ever change the weight of the `initialSecurityModule` (and `l1TotalWeight)`?

Reply:
Just opened an issue on `teleporter` about that: https://github.com/ava-labs/teleporter/issues/645

For the meantime, maybe we can just remove `l1TotalWeight` as we cannot update it properly.